SSL sign up issues

While an SSL certificate is present on the site (thanks Augusto in #102 for checking), HTTP is not redirected to HTTPS as it should. Furthermore, the site is shown as not completely secure because the absolute link to the newsletter sign-up is to an HTTP resources. We should probably just point this to the GitHub issues and appropriate Discourse forum.

The specific error in the dev console is:

```Mixed Content: The page at 'https://publicbodies.org/' was loaded over a secure connection, but contains a form that targets an insecure endpoint 'http://lists.okfn.org/mailman/subscribe/okfn-labs'. This endpoint should be made available over a secure connection.```

I have replaced the HTTP link on that old form in PR #108.

For the HTTP -> HTTPS redirection to work, however, it would be necessary to change the application code. But this is ancient node.js, and before changing it it would be advisable to upgrade it to a more recent node.js version. That also means updating / getting rid of all those deprecated and obsolete dependencies:

```
npm WARN deprecated less-middleware@0.1.15: Updated API and improved functionality.
npm WARN deprecated mkdirp@0.3.5: Legacy versions of mkdirp are no longer supported. Please update to mkdirp 1.x. (Note that the API surface has changed to use Promises in 1.x.)
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported

> leveldown@6.0.0 install /home/node/portal/node_modules/leveldown
> node-gyp-build

npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN publicbodies@0.1.0 license should be a valid SPDX license expression

added 159 packages from 149 contributors and audited 160 packages in 11.598s

13 packages are looking for funding
  run `npm fund` for details

found 29 vulnerabilities (10 low, 11 moderate, 7 high, 1 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
```

I tried to do it, but then got an error:

```
internal/modules/cjs/loader.js:818
  throw err;
  ^

Error: Cannot find module '/home/node/portal/express'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)
    at Function.Module._load (internal/modules/cjs/loader.js:667:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

which is strange, as the `express` package *is* installed.

I suspect that to modernize Javascript and node.js on this application would require rewriting a good chunk of it and would also be beyond my weak abilities in this area.

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,18 +1,20 @@
 # DESCRIPTION: publicbodies.org website
 # BUILD: docker build --rm -t publicbodies .
 # 
-# If you want to just build the documentation portal locally, do:
-#   docker build --rm -t publicbodies .
+# If you want to build the portal locally in development mode, do:
+#   docker build --rm -t publicbodies . --build-arg NODE_ENV=development
 # to build the container, then run
 #   docker run --rm --name publicbodies -p 3000:3000 -it publicbodies node index.js
 # open http://localhost:3000 on your browser to see the portal.
 
 FROM node:12
 
+ARG NODE_ENV=production
+
 # Never prompt the user for choices on installation/configuration of packages
 ENV DEBIAN_FRONTEND noninteractive
 ENV TERM linux
-ENV NODE_ENV=production
+ENV NODE_ENV=$NODE_ENV
 # from https://github.com/nodejs/docker-node/blob/main/docs/BestPractices.md#global-npm-dependencies
 ENV NPM_CONFIG_PREFIX=/home/node/.npm-global
 ENV PATH=$PATH:/home/node/.npm-global/bin

--- a/Dockerfile
+++ b/Dockerfile
@@ -4,7 +4,7 @@
 # If you want to just build the documentation portal locally, do:
 #   docker build --rm -t publicbodies .
 # to build the container, then run
-#   docker run --rm --volume="$PWD:/home/node/portal" -p 3000:3000 -it publicbodies node index.js
+#   docker run --rm --name publicbodies -p 3000:3000 -it publicbodies node index.js
 # open http://localhost:3000 on your browser to see the portal.
 
 FROM node:12
@@ -25,6 +25,14 @@ RUN mkdir -p /home/node/portal
 COPY ./package.json /home/node/portal/package.json
 WORKDIR /home/node/portal
 RUN npm install .
+
+# copy website resources
+COPY ./lib /home/node/portal/lib
+COPY ./routes /home/node/portal/routes
+COPY ./views /home/node/portal/views
+COPY ./public /home/node/portal/public
+COPY ./data /home/node/portal/data
+COPY ./index.js /home/node/portal/
 
 # port Docusaurus runs on
 EXPOSE 3000

--- a/README.md
+++ b/README.md
@@ -61,7 +61,7 @@ The website is a node webapp. To get it running:
   If you're using Docker, start the container instead:
 
         ```
-        docker run --rm --volume="$PWD:/home/node/portal" -p 3000:3000 -it publicbodies node index.js
+        docker run --rm --name publicbodies -p 3000:3000 -it publicbodies node index.js
         ```
 
 The list of outstanding issues is at: <https://github.com/okfn/publicbodies/issues>

--- a/README.md
+++ b/README.md
@@ -51,6 +51,13 @@ The website is a node webapp. To get it running:
         ```
         docker build --rm -t publicbodies .
         ```
+  
+  If you are building a development environment, please use:
+
+        ```
+        docker build --rm -t publicbodies . --build-arg NODE_ENV=development
+        ```
+  so that you can get debugging information.
 
 4. Run the webapp:
 

--- a/index.js
+++ b/index.js
@@ -3,6 +3,7 @@
  */
 
 var express = require('express'),
+    errorhandler = require('errorhandler'),
     favicon = require('serve-favicon'),
     morgan = require('morgan'),
     lessMiddleware = require('less-middleware'),
@@ -22,8 +23,9 @@ app.use(lessMiddleware(__dirname + '/public'));
 app.use(express.static(path.join(__dirname, 'public')));
 
 // development only
-if ('development' === app.get('env')) {
-    app.use(express.errorHandler());
+if ('development' === process.env.NODE_ENV) {
+    app.use(errorhandler());
+}
 }
 
 app.get('/', routes.index);

--- a/index.js
+++ b/index.js
@@ -26,6 +26,16 @@ app.use(express.static(path.join(__dirname, 'public')));
 if ('development' === process.env.NODE_ENV) {
     app.use(errorhandler());
 }
+else { // production only
+    // handle https redirect
+    function ensureSecure(req, res, next){
+        if(req.secure){
+        // OK, continue
+        return next();
+        };
+        res.redirect('https://' + req.hostname + req.url);
+    }
+    app.all('*', ensureSecure);
 }
 
 app.get('/', routes.index);

--- a/package.json
+++ b/package.json
@@ -33,6 +33,7 @@
     "async": "~0.2.9",
     "csv": "~5.0.0",
     "dirty": "~0.9.7",
+    "errorhandler": "~1.5.1",
     "express": "^4.7.4",
     "less-middleware": "~3.1.0",
     "morgan": "~1.10.0",

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "publicbodies",
-  "version": "0.1.0",
+  "version": "0.1.1",
   "description": "A database of public bodies such as government departments, ministries etc.",
   "main": "site/make.js",
   "directories": {

